### PR TITLE
ACL group attach/detach to ENI

### PR DESCRIPTION
**Pull request by mukeshmv**
1) Support for 5 stage v4 and v6 ACL group IDs driven from ENI table as specified in the GNMI schema.
2) New ACL group table and change name of ACL table to ACL_RULE table.
3) Change P4 key match type to "optional" to trigger generation of Priority attribute in SAI API.

Adds/removes the following SAI attributes

```
New ACL group table -
/**
 * @brief Attribute ID for dash_acl_dash_acl_group
 */
typedef enum _sai_dash_acl_group_attr_t
{

    SAI_DASH_ACL_GROUP_ATTR_START,
    /**
     * @brief Action set_acl_group_attrs parameter IP_ADDR_FAMILY
     * @type sai_ip_addr_family_t
     * @flags CREATE_AND_SET
     * @default SAI_IP_ADDR_FAMILY_IPV4
     */
    SAI_DASH_ACL_GROUP_ATTR_IP_ADDR_FAMILY = SAI_DASH_ACL_GROUP_ATTR_START,
    SAI_DASH_ACL_GROUP_ATTR_END,
    SAI_DASH_ACL_GROUP_ATTR_CUSTOM_RANGE_START = 0x10000000,
    SAI_DASH_ACL_GROUP_ATTR_CUSTOM_RANGE_END,

} sai_dash_acl_group_attr_t;

typedef sai_status_t (*sai_create_dash_acl_group_fn)(
        _Out_ sai_object_id_t *dash_acl_group_id,
        _In_ sai_object_id_t switch_id,
        _In_ uint32_t attr_count,
        _In_ const sai_attribute_t *attr_list);

Change ACL attributes - 
typedef enum _sai_dash_acl_rule_attr_t
{
     ...
    /* Added */ 
    SAI_DASH_ACL_RULE_ATTR_ACL_GROUP_ID,
    /* Removed */
    SAI_DASH_ACL_ATTR_ACL_STAGE_ID,
    ...
}  sai_dash_acl_attr_t;

New ENI attributes to drive ACL group
/**
 * @brief Attribute ID for dash_eni
 */
typedef enum _sai_eni_attr_t
{
    ...

    /* Added */ 

    SAI_ENI_ATTR_INBOUND_V4_STAGE1_ACL_GROUP_ID,
    SAI_ENI_ATTR_INBOUND_V4_STAGE2_ACL_GROUP_ID,
    SAI_ENI_ATTR_INBOUND_V4_STAGE3_ACL_GROUP_ID,
    SAI_ENI_ATTR_INBOUND_V4_STAGE4_ACL_GROUP_ID,
    SAI_ENI_ATTR_INBOUND_V4_STAGE5_ACL_GROUP_ID,

    SAI_ENI_ATTR_INBOUND_V6_STAGE1_ACL_GROUP_ID,
    SAI_ENI_ATTR_INBOUND_V6_STAGE2_ACL_GROUP_ID,
    SAI_ENI_ATTR_INBOUND_V6_STAGE3_ACL_GROUP_ID,
    SAI_ENI_ATTR_INBOUND_V6_STAGE4_ACL_GROUP_ID,
    SAI_ENI_ATTR_INBOUND_V6_STAGE5_ACL_GROUP_ID,

    SAI_ENI_ATTR_OUTBOUND_V4_STAGE1_ACL_GROUP_ID,
    SAI_ENI_ATTR_OUTBOUND_V4_STAGE2_ACL_GROUP_ID,
    SAI_ENI_ATTR_OUTBOUND_V4_STAGE3_ACL_GROUP_ID,
    SAI_ENI_ATTR_OUTBOUND_V4_STAGE4_ACL_GROUP_ID,
    SAI_ENI_ATTR_OUTBOUND_V4_STAGE5_ACL_GROUP_ID,

    SAI_ENI_ATTR_OUTBOUND_V6_STAGE1_ACL_GROUP_ID,
    SAI_ENI_ATTR_OUTBOUND_V6_STAGE2_ACL_GROUP_ID,
    SAI_ENI_ATTR_OUTBOUND_V6_STAGE3_ACL_GROUP_ID,
    SAI_ENI_ATTR_OUTBOUND_V6_STAGE4_ACL_GROUP_ID,
    SAI_ENI_ATTR_OUTBOUND_V6_STAGE5_ACL_GROUP_ID,

    ...
} sai_eni_attr_t;

### CODE REVIEW

**Comment by mukeshmv** [comment]
@marian-pritsak I have removed the changes to the SAI implementation template and the vnet_out test from this PR. Please check.

File: dash-pipeline/SAI/sai_api_gen.py [hunk 1]
```diff
@@ -118,6 +120,11 @@ def get_sai_key_data(key):
     else:
         raise ValueError(f"match_type={sai_key_data['match_type']} is not supported")
 
+    if 'sai_key_type' in sai_key_data and  sai_key_data['sai_key_type'] == 'sai_object_id_t':
+        if sai_key_data['sai_key_name'].endswith('dash_acl_group_id'):
```
Comment (marian-pritsak) [inline dash-pipeline/SAI/sai_api_gen.py:124]: Why is there an explicit reference to a table from DASH?

File: dash-pipeline/SAI/templates/saiapi.h.j2 [hunk 2]
```diff
@@ -219,7 +199,7 @@ typedef enum _sai_{{ table.name }}_attr_t
     /**
      * @brief Rule priority in table
      *
-     * @type sai_uint32_t
+     * @type sai_int32_t
```
Comment (marian-pritsak) [inline dash-pipeline/SAI/templates/saiapi.h.j2:202]: Why? Priority can't be negative

**Review by marian-pritsak — CHANGES_REQUESTED** [review]

File: dash-pipeline/SAI/templates/saiapi.h.j2 (hunk 2 above)
Comment (mukeshmv) [reply dash-pipeline/SAI/templates/saiapi.h.j2:202]: Yes will remove this change. I think I went by the APP_DB schema which had it as INT32
`priority                 = INT32 value  ; priority of the rule, lower the value, higher the priority`

File: dash-pipeline/SAI/sai_api_gen.py (hunk 1 above)
Comment (mukeshmv) [reply dash-pipeline/SAI/sai_api_gen.py:124]: Will remove this. My original change had the acl rule key name as acl_group_id. So I had added this to fix the @objects meta in the header to the correct dash table name. But later I changed the acl rule key name to dash_acl_group_id. So its not needed here anymore. 

Now this is only needed for ENI params like SAI_ENI_ATTR_INBOUND_V4_STAGE1_DASH_ACL_GROUP_ID.

Comment (mukeshmv) [reply dash-pipeline/SAI/sai_api_gen.py:124]: Will remove this. My original change had the acl rule key name as acl_group_id. So I had added this to fix the @objects meta in the header to the correct dash table name. But later I changed the acl rule key name to dash_acl_group_id. So its not needed here anymore.

File: dash-pipeline/SAI/sai_api_gen.py [hunk 3]
```diff
@@ -136,6 +138,11 @@ def extract_action_data(program):
                 param[NAME_TAG] = p[NAME_TAG]
                 param['type'], param['field'] = get_sai_key_type(int(p[BITWIDTH_TAG]), p[NAME_TAG], p[NAME_TAG])
                 param['bitwidth'] = p[BITWIDTH_TAG]
+                if param['type'] == 'sai_object_id_t':
+                    if param[NAME_TAG].endswith('dash_acl_group_id'):
```
Comment (marian-pritsak) [inline dash-pipeline/SAI/sai_api_gen.py:142]: There shouldn't be a reference to a table from DASH pipeline. It should be deducted from the list of tables after the API dictionary is fully formed by going over the tables once again, kind of like that:
```
for param in params:
    for table in tables:
        if param[NAME_TAG].endswith(table.[NAME_TAG]):
            param[OBJECT_NAME_TAG] = table[NAME_TAG][:-len("_id")]
```

File: dash-pipeline/bmv2/dash_pipeline.p4 [hunk 4]
```diff
@@ -83,18 +83,50 @@ control dash_ingress(inout headers_t hdr,
...
+   meta.stage5_dash_acl_group_id = ## prefix ##_stage5_dash_acl_group_id;
+
     action set_eni_attrs(bit<32> cps,
                          bit<32> pps,
                          bit<32> flows,
                          bit<1> admin_state,
                          IPv4Address vm_underlay_dip,
-                         bit<24> vm_vni) {
+                         bit<24> vm_vni,
+                         ACL_GROUPS_PARAM(inbound_v4),
```
Comment (marian-pritsak) [inline dash-pipeline/bmv2/dash_pipeline.p4:106]: There should also be a differentiation between inbound and outbound group IDs. Inbound v4 ACL stage 1 != Outbound v4 ACL stage 1.

**Review by marian-pritsak — CHANGES_REQUESTED** [review]

File: dash-pipeline/SAI/sai_api_gen.py (hunk 3 above)
Comment (mukeshmv) [reply dash-pipeline/SAI/sai_api_gen.py]: Fixed

**Review by marian-pritsak — APPROVED** [review]

**Comment by marian-pritsak** [comment]
Looks cool.

**Pull request merged**